// lamington/lib/cli/logIndicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ora_1 = require("ora");
const colors = require("colors");
/** Holds spinner instances */
const cache = {};
/**
 * Creates a new spinner instance with the specified message
 * @author Mitch Pierias <github.com/MitchPierias>
 * @param text Output display message
 */
exports.create = (text) => {
    // Cleanup existing spinner
    if (cache.spinner) {
        cache.spinner.succeed();
        delete cache.spinner;
    }
    // Create and cache spinner
    exports.update(text);
};
/**
 * Updates the text of the current spinner
 * @author Mitch Pierias <github.com/MitchPierias>
 * @param text CLI output text to display
 */
exports.update = (text) => {
    // Create spinner when missing
    if (!cache.spinner)
        cache.spinner = ora_1.default({
            text: colors.white(text),
            color: 'magenta',
        }).start();
    // Update spinner text
    cache.spinner.text = colors.white(text);
};
/**
 * Terminates the current spinner with the specified output message
 * @author Mitch Pierias <github.com/MitchPierias>
 * @param message Output message
 * @param isError Renders output as error toggle
 */
exports.end = (message = '', isError = false) => {
    // Check spinner reference
    if (!cache.spinner)
        return;
    // Handle output
    if (isError) {
        cache.spinner.fail(colors.grey(message));
    }
    else {
        cache.spinner.succeed(colors.grey(message));
    }
    // Clear spinner reference
    delete cache.spinner;
};
/**
 * Terminates the current spinner as an error with output message
 * @author Mitch Pierias <github.com/MitchPierias>
 * @param message Spinner message.
 */
exports.fail = (message) => {
    exports.end(message, true);
};
